src/station_cmd.cpp

```diff
@@ -4014,68 +4014,108 @@ const StationList *StationFinder::GetStations()
 
-		/* This station can be used, add it to st1/st2 */
-		if (st1 == nullptr || st->goods[type].rating >= best_rating1) {
-			st2 = st1; best_rating2 = best_rating1; st1 = st; best_rating1 = st->goods[type].rating;
-		} else if (st2 == nullptr || st->goods[type].rating >= best_rating2) {
-			st2 = st; best_rating2 = st->goods[type].rating;
-		}
-	}
-
-	/* no stations around at all? */
-	if (st1 == nullptr) return 0;
+	/* Lowest possible rating, better not to give cargo anymore. */
+	if (st->goods[type].rating == 0) return false;
 
-	/* From now we'll calculate with fractal cargo amounts.
-	 * First determine how much cargo we really have. */
-	amount *= best_rating1 + 1;
+	/* Selectively servicing stations, and not this one. */
+	if (_settings_game.order.selectgoods && !st->goods[type].HasVehicleEverTriedLoading()) return false;
 
-	if (st2 == nullptr) {
-		/* only one station around */
-		return UpdateStationWaiting(st1, type, amount, source_type, source_id);
+	if (IsCargoInClass(type, CC_PASSENGERS)) {
+		/* Passengers are never served by just a truck stop. */
+		if (st->facilities == FACIL_TRUCK_STOP) return false;
+	} else {
+		/* Non-passengers are never served by just a bus stop. */
+		if (st->facilities == FACIL_BUS_STOP) return false;
 	}
+	return true;
+}
 
-	/* several stations around, the best two (highest rating) are in st1 and st2 */
-	assert(st1 != nullptr);
-	assert(st2 != nullptr);
-	assert(best_rating1 != 0 || best_rating2 != 0);
-
-	/* Then determine the amount the worst station gets. We do it this way as the
-	 * best should get a bonus, which in this case is the rounding difference from
-	 * this calculation. In reality that will mean the bonus will be pretty low.
-	 * Nevertheless, the best station should always get the most cargo regardless
-	 * of rounding issues. */
-	uint worst_cargo = amount * best_rating2 / (best_rating1 + best_rating2);
-	assert(worst_cargo <= (amount - worst_cargo));
-
-	/* And then send the cargo to the stations! */
-	uint moved = UpdateStationWaiting(st1, type, amount - worst_cargo, source_type, source_id);
-	/* These two UpdateStationWaiting's can't be in the statement as then the order
-	 * of execution would be undefined and that could cause desyncs with callbacks. */
-	return moved + UpdateStationWaiting(st2, type, worst_cargo, source_type, source_id);
+uint MoveGoodsToStation(CargoID type, uint amount, SourceType source_type, SourceID source_id, const StationList *all_stations)
+{
+    /* Return if nothing to do. Also the rounding below fails for 0. */
+    if (all_stations->size() == 0) return 0;
+    if (amount == 0) return 0;
+
+    Station *first_station = nullptr;
+    typedef std::pair<Station *, uint> StationInfo;
+    std::vector<StationInfo> used_stations;
+
+    for (Station *st : *all_stations) {
+        if (!CanMoveGoodsToStation(st, type)) continue;
+
+        /* Avoid allocating a vector if there is only one station to significantly
```

Comment:
Can you provide some numbers on this optimisation? I feel like a `.reserve(1)` with the declaration of `used_stations` would have the same result, and have significantly simpler code

Reply:
reserve(1) is about 3x slower for 1 station on gcc. Should be similar on other compilers as empty vectors are more likely to be optimized than vectors of size 1.